lib/spack/spack/spec.py

```diff
@@ -543,64 +544,29 @@ class CompilerSpec(object):
 
     __slots__ = "name", "versions"
 
-    def __init__(self, *args):
-        nargs = len(args)
-        if nargs == 1:
-            arg = args[0]
-            # If there is one argument, it's either another CompilerSpec
-            # to copy or a string to parse
-            if isinstance(arg, str):
-                spec = spack.parser.parse_one_or_raise(f"%{arg}")
-                self.name = spec.compiler.name
-                self.versions = spec.compiler.versions
-
-            elif isinstance(arg, CompilerSpec):
-                self.name = arg.name
-                self.versions = arg.versions.copy()
-
-            else:
-                raise TypeError(
-                    "Can only build CompilerSpec from string or "
-                    + "CompilerSpec. Found %s" % type(arg)
-                )
-
-        elif nargs == 2:
-            name, version = args
-            self.name = name
-            self.versions = vn.VersionList()
-            versions = vn.ver(version)
-            self.versions.add(versions)
-
-        else:
-            raise TypeError("__init__ takes 1 or 2 arguments. (%d given)" % nargs)
-
-    def _add_versions(self, version_list):
-        # If it already has a non-trivial version list, this is an error
-        if self.versions and self.versions != vn.VersionList(":"):
-            # Note: This may be impossible to reach by the current parser
-            # Keeping it in case the implementation changes.
-            raise MultipleVersionError(
-                "A spec cannot contain multiple version signifiers." " Use a version list instead."
-            )
+    def __init__(self, name: str, version: str):
```

Comment:
I think the usage of this was way more clear before -- why is it bad for the constructor to take multiple types?

I think looking at the `@overload` example here: https://mypy.readthedocs.io/en/stable/more_types.html might help (`@overload` is supported in 3.6).  I think we could do a less invasive (and more readable) cleanup using that pattern.

Reply:
I didn't know `@overload`, so I'll try that and push the change.

A little explanation of the choice here. Modulo the name of the function (`maybe_literal`, which I'm definitely not tied to) using `*args` like this seems a lot like using `**kwargs` in other places. For instance, looking at this code yesterday I noticed that:
```
CompilerSpec("gcc")
```
and
```
CompilerSpec("gcc", ":")
```
were taking two different code paths with the first one doing strictly more work since it goes through parsing the string. Also, it seems taking a `CompilerSpec` in the constructor is dead code, since we always use `_autospec` in all the places where we can potentially have an object of the same type. Finally, we can't name the arguments in the current state like:
```
CompilerSpec("gcc", version=":")
```
since the constructor is taking `*args`. So, none of the above is a major concern, but I slightly preferred this version because it's clear which code path we are taking from looking at the call site.